spec.html

```diff
@@ -30110,19 +30110,59 @@ <h1>Function.prototype.toString ( )</h1>
             1. Return CodePointsToString(_func_.[[SourceText]]).
-          1. If _func_ is a <emu-xref href="#sec-built-in-function-objects">built-in function object</emu-xref>, return an implementation-defined String source code representation of _func_. The representation must have the syntax of a |NativeFunction|. Additionally, if _func_ has an [[InitialName]] internal slot and _func_.[[InitialName]] is a String, the portion of the returned String that would be matched by |NativeFunctionAccessor?| |PropertyName| must be the value of _func_.[[InitialName]].
-          1. If Type(_func_) is Object and IsCallable(_func_) is *true*, return an implementation-defined String source code representation of _func_. The representation must have the syntax of a |NativeFunction|.
-          1. Throw a *TypeError* exception.
+          1. Else if _func_ is a <emu-xref href="#sec-built-in-function-objects">built-in function object</emu-xref>, then
+            1. Let _sourceText_ be an implementation-defined sequence of Unicode code points that satisfies the syntax of |NativeFunction| and represents _func_.
+          1. Else if Type(_func_) is Object and IsCallable(_func_) is *true*, then
+            1. Let _sourceText_ be an implementation-defined sequence of Unicode code points that satisfies the syntax of |NativeFunction| and represents _func_.
+          1. Else,
+            1. Throw a *TypeError* exception.
+          1. Assert: ParseText(_sourceText_, |NativeFunction|) is a Parse Node.
+          1. If _func_ has an [[InitialName]] internal slot and _func_.[[InitialName]] is a String, then
+            1. Assert: NativeFunctionName(_sourceText_) is _func_.[[InitialName]].
+          1. Return CodePointsToString(_sourceText_).
         </emu-alg>
 
         <emu-grammar type="definition">
           NativeFunction :
-            `function` NativeFunctionAccessor? PropertyName[~Yield, ~Await]? `(` FormalParameters[~Yield, ~Await] `)` `{` `[` `native` `code` `]` `}`
+            `function` NativeFunctionName? `(` FormalParameters[~Yield, ~Await] `)` `{` `[` `native` `code` `]` `}`
+
+          NativeFunctionName :
+            NativeFunctionAccessor? LiteralPropertyName
+            NativeFunctionAccessor? `[` StringLiteral `]`
 
           NativeFunctionAccessor :
             `get`
             `set`
         </emu-grammar>
+
+        <emu-clause id="sec-nativefunctionname" type="abstract operation">
+          <h1>
+            NativeFunctionName (
+              _sourceText_: a sequence of Unicode code points,
+            ): a String
+          </h1>
+          <dl class="header">
+            <dt>description</dt>
+            <dd>It returns the name of a built-in function object represented by source text.</dd>
+          </dl>
+          <emu-alg>
+            1. Let _function_ be ParseText(_sourceText_, |NativeFunction|).
+            1. If _name_ Contains |NativeFunctionName| is *false*, return *""*.
```

Comment:
`_name_` isn't introduced yet here. Also, don't use Contains for this since `FormalParameters` can contain arbitrary expressions. You can say something like "if `|NativeFunctionName|` is present in `_function_`".

Reply:
Good point, I've updated accordingly.